### PR TITLE
applications: asset_tracker: add sensor threshold cfg

### PR DESCRIPTION
**Pull request by jayteemo**
Add high/low threshold config command handling for sensor data.

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

### CODE REVIEW

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 1]
```diff
@@ -195,10 +230,35 @@ static const char *const channel_type_str[] = {
...
+	{ .ch = CLOUD_CHANNEL_LIGHT_BLUE,
+	  .cfg = { .value = { [SENSOR_CH_CFG_ITEM_TYPE_SEND_ENABLE] = true } } },
+	{ .ch = CLOUD_CHANNEL_LIGHT_IR,
+	  .cfg = { .value = { [SENSOR_CH_CFG_ITEM_TYPE_SEND_ENABLE] = true } } }
+};
 static cloud_cmd_cb_t cloud_command_cb;
 struct cloud_command cmd_parsed;
 
+static int
+cloud_cmd_handle_sensor_set_ch_cfg(struct cloud_command const *const cmd);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: I think convention is to instead break the prototype up at the parameters if it exceed 80 chars, but personally no strong feelings about it and can't find anything in the code style about it

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 2]
```diff
@@ -195,10 +230,35 @@ static const char *const channel_type_str[] = {
...
+
+static int
+cloud_set_ch_cfg_item(const enum cloud_channel channel,
+			  const enum sensor_ch_cfg_item_type type,
+			  const double value);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Either something up with the alignment, or GitHub acting up again

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 3]
```diff
@@ -367,6 +426,31 @@ static int cloud_search_cmd(cJSON *group_obj, enum cloud_cmd_group group)
...
+						type_obj->valuedouble;
+				} else {
+					continue;
+				}
+
+				if ((group == CLOUD_CMD_GROUP_SET) &&
+				    (cloud_cmd_handle_sensor_set_ch_cfg(
+					     &cmd_parsed) == 0)) {
+					/* no need to pass to cb if */
+					/* cmd was successfully handed */
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: typo, `handled`

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 4]
```diff
@@ -391,24 +475,25 @@ int cloud_decode_command(char const *input)
...
+		return -ENOTSUP;
+	}
+
+	if (get_obj != NULL) {
 		cmd_parsed.group = group_get.group;
-		ret = cloud_search_cmd(group_obj, cmd_parsed.group);
-		if (ret) {
-			return ret;
-		}
+		(void)cloud_search_cmd(get_obj, cmd_parsed.group);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Should the returned error code be checked?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 5]
```diff
@@ -461,19 +546,39 @@ int cloud_encode_env_sensors_data(const env_sensor_data_t *sensor_data,
 #if CONFIG_LIGHT_SENSOR
 /* 4 32-bit ints, 3 spaces, NULL */
 #define LIGHT_SENSOR_DATA_STRING_MAX_LEN ((4 * 11) + 3 + 1)
+#define LIGHT_SENSOR_DATA_NO_UPDATE (-1)
 int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Missing blank line before function

Comment (jtguggedal) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Not subject, of this PR, but looks like pointer could also be `const`

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 6]
```diff
@@ -461,19 +546,39 @@ int cloud_encode_env_sensors_data(const env_sensor_data_t *sensor_data,
...
 	if ((sensor_data == NULL) || (output == NULL)) {
 		return -EINVAL;
 	}
 
-	len = snprintf(buf, sizeof(buf), "%d %d %d %d", sensor_data->red,
-		       sensor_data->green, sensor_data->blue, sensor_data->ir);
+	if (cloud_is_send_allowed(CLOUD_CHANNEL_LIGHT_RED, sensor_data->red)) {
+		send.red = sensor_data->red;
+	}
+	if (cloud_is_send_allowed(CLOUD_CHANNEL_LIGHT_GREEN,
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Some blank lines missing between the `if`s

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 7]
```diff
@@ -482,3 +587,124 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
 	return cloud_encode_data(&cloud_sensor, output);
 }
 #endif /* CONFIG_LIGHT_SENSOR */
+
+static int sensor_ch_cfg_set_item(struct sensor_ch_cfg *const cfg,
+				  const enum sensor_ch_cfg_item_type type,
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Some alignment issues here

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 8]
```diff
@@ -482,3 +587,124 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
...
+	    (type >= SENSOR_CH_CFG_ITEM_TYPE__END) || (cfg == NULL)) {
+		return -EINVAL;
+	}
+
+	cfg->value[type] = value;
+
+	return 0;
+}
+
+static bool sensor_ch_cfg_is_send_allowed(const struct sensor_ch_cfg *const cfg,
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Function name reads like a question to me. Not necessarily a bad thing, though, just not used to it :)

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 9]
```diff
@@ -482,3 +587,124 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
...
+	     (sensor_value >
+	      cfg->value[SENSOR_CH_CFG_ITEM_TYPE_THRESH_HIGH_VALUE]))) {
+		return true;
+	}
+
+	return (!cfg->value[SENSOR_CH_CFG_ITEM_TYPE_THRESH_LOW_ENABLE] &&
+		!cfg->value[SENSOR_CH_CFG_ITEM_TYPE_THRESH_HIGH_ENABLE]);
+}
+
+int cloud_set_ch_cfg_item(const enum cloud_channel channel,
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: `static`?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 10]
```diff
@@ -482,3 +587,124 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
...
+		if (sensor_cfg[i].ch == channel) {
+			return sensor_ch_cfg_is_send_allowed(&sensor_cfg[i].cfg,
+							     value);
+		}
+	}
+
+	return false;
+}
+
+int cloud_cmd_handle_sensor_set_ch_cfg(struct cloud_command const *const cmd)
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: `static`?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h [hunk 11]
```diff
@@ -113,13 +121,20 @@ enum cloud_cmd_type {
...
+
 struct cloud_command {
 	enum cloud_cmd_group group; /* The group the decoded command belongs to. */
 	enum cloud_cmd_recipient recipient; /* The command's recipient module. */
 	enum cloud_channel channel; /* The command's desired channel. */
 	enum cloud_cmd_type type; /* The command type, the desired action. */
 	double value; /* The value to be written to the recipient/channel. */
-	bool state; /* The truth value to be written to the recipient/channel. */
+	enum cloud_cmd_state
+		state; /* The truth value to be written to the recipient/channel. */
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.h]: Let's put it on the same line and have comment on the line before

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h [hunk 12]
```diff
@@ -186,6 +201,17 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
...
 #endif /* CONFIG_LIGHT_SENSOR */
 
+
+/**
+ * @brief Checks if data should be sent to the cloud based on config.
+ *
+ * @param channel The cloud channel type..
+ * @param value Current data value for channel.
+ *
+ * @return true if the data should be sent to the cloud.
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.h]: Capital letter in `If`

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h [hunk 13]
```diff
@@ -186,6 +201,17 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
 				   struct cloud_msg *output);
 #endif /* CONFIG_LIGHT_SENSOR */
 
+
+/**
+ * @brief Checks if data should be sent to the cloud based on config.
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.h]: Should or could?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 14]
```diff
@@ -540,7 +540,7 @@ static void nct_mqtt_evt_handler(struct mqtt_client *const mqtt_client,
 	case MQTT_EVT_PUBLISH: {
 		const struct mqtt_publish_param *p = &_mqtt_evt->param.publish;
 
-		LOG_DBG("MQTT_EVT_PUBLISH: id=%d len=%d ",
+		LOG_DBG("MQTT_EVT_PUBLISH: id=%d len=%d",
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c]: Spaces around `=` would be nice

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h (hunk 13 above)
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.h]: `should`  but in the context of the fuction name `could` is probably best.

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 4 above)
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: it did not seem necessary to me.  it is just searching for a potential command.  current error sources are a null parameter or bad enum value.  i could add an error code check and print an error msg?

it definitely should not return immediately though, as the parsed json obj needs to be deleted.

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 1 above)
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: i believe this was clang-format's doing 🙃

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 15]
```diff
@@ -31,6 +31,27 @@ struct cmd {
...
+	SENSOR_CH_CFG_ITEM_TYPE__END
+};
+
+struct sensor_ch_cfg {
+	double value[SENSOR_CH_CFG_ITEM_TYPE__END];
+};
+
+struct cloud_sensor_ch_cfg {
+	enum cloud_channel ch;
+	struct sensor_ch_cfg cfg;
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: i have no preference, i can change it if you think it is valuable.
the items in question are local to this file though, not a shared header.

[the remark this answers is not in the document]
Comment (jtguggedal) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Okay, let's go for then `chan` for consistency across the codec files

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 4 above)
Comment (jtguggedal) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Right, seems reasonable to ignore it

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h (hunk 13 above)
Comment (jtguggedal) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.h]: You're the native English speaker in the room, so you decide :)

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 16]
```diff
@@ -195,10 +230,34 @@ static const char *const channel_type_str[] = {
...
+	{ .chan = CLOUD_CHANNEL_LIGHT_BLUE,
+	  .cfg = { .value = { [SENSOR_CHAN_CFG_ITEM_TYPE_SEND_ENABLE] = true } } },
+	{ .chan = CLOUD_CHANNEL_LIGHT_IR,
+	  .cfg = { .value = { [SENSOR_CHAN_CFG_ITEM_TYPE_SEND_ENABLE] = true } } }
+};
 static cloud_cmd_cb_t cloud_command_cb;
 struct cloud_command cmd_parsed;
 
+static int cloud_cmd_handle_sensor_set_chan_cfg(struct cloud_command const
+		      *const cmd);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Alignment looks a bit off

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 17]
```diff
@@ -195,10 +230,34 @@ static const char *const channel_type_str[] = {
...
+
+static int cloud_set_chan_cfg_item(const enum cloud_channel channel,
+		      const enum sensor_chan_cfg_item_type type,
+			  const double value);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Some alignment things going on here as well

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 18]
```diff
@@ -195,10 +230,34 @@ static const char *const channel_type_str[] = {
 	[CLOUD_CHANNEL_DEVICE_INFO] = CLOUD_CHANNEL_STR_DEVICE_INFO,
 	[CLOUD_CHANNEL_LIGHT_SENSOR] = CLOUD_CHANNEL_STR_LIGHT_SENSOR,
 };
-
+struct cloud_sensor_chan_cfg sensor_cfg[] = {
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: `static`?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 6 above)
Comment (jtguggedal) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Little bit surprised checkpatch doesn't complain about missing blank line before/after if, but as long as it passes I guess it's okay wrt NCS code style

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 19]
```diff
@@ -508,3 +612,124 @@ int cloud_encode_light_sensor_data(const struct light_sensor_data *sensor_data,
...
+		return -EINVAL;
+	}
+
+	cfg->value[type] = value;
+
+	return 0;
+}
+
+static bool sensor_chan_cfg_is_send_allowed(const struct sensor_chan_cfg *const cfg,
+					  const double sensor_value)
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Alignment issue?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 20]
```diff
@@ -570,7 +570,7 @@ static void nct_mqtt_evt_handler(struct mqtt_client *const mqtt_client,
 		} else {
 			/* Try to match it with one of the data topics. */
 			dc.id = p->message_id;
-			dc.data.ptr = p->message.payload.data;
+			dc.data.ptr = nct.payload_buf;
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c]: Nice catch

Comment (rlubos) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c]: If it's a bugfix, it should be in a separate commit.

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 6 above)
Comment (rlubos) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: I agree here, personally I prefer to separate such expressions with an empty line, but as it doesn't seem to be enforced by the conding standard, I just let it be.

**Comment by jayteemo** [comment]
@rlubos any idea why the builder is stuck?

**Comment by rlubos** [comment]
@jayteemo For whatever reason CI does not recognize you as a Nordic employee...

@thst-nordic Can you look at this issue and add @jayteemo to any group needed?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 21]
```diff
@@ -195,10 +230,34 @@ static const char *const channel_type_str[] = {
...
+};
 static cloud_cmd_cb_t cloud_command_cb;
 struct cloud_command cmd_parsed;
 
+static int cloud_cmd_handle_sensor_set_chan_cfg(struct cloud_command const
+		      	  	  	  	  	  	  	  	  	*const cmd);
+
+static int cloud_set_chan_cfg_item(const enum cloud_channel channel,
+		      	  	  	  	  	   const enum sensor_chan_cfg_item_type type,
+								   const double value);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Still weird things going on with the formatting, but could be for me locally on github

Comment (rlubos) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: This actually looks off and should be fixed. The CI actually did not run over this PR yet, I hope that `checkpatch` will notice such issues.

Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: this is the format given by both clang-format and uncrustify...
for line 254 it gives:
```static int
cloud_cmd_handle_sensor_set_chan_cfg(struct cloud_command const *const cmd);```

Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: i have no preference on these matters.  i would like to find a tool that produces the desired output, to save everyone time.

**Pull request merged**